File: iosApp/iosApp/UI/Screens/Group/GroupEventsViewModel.swift
import Combine
import Foundation
import Shared

/// 일정 탭 — composeApp GroupEventsViewModel.kt와 1:1 미러(웹 /groups/[id]/events 캘린더 페이지 미러,
/// 탭별 VM 분리). 월 앵커 기준 [월초, 다음달 초) 범위를 조회하고, 날짜 귀속·선택 키는 기기 로컬
/// 타임존 yyyy-MM-dd(웹 ymd 미러). RSVP는 낙관적 갱신 없이 서버가 돌려준 집계 갱신 일정으로 카드를
/// 교체하고, 같은 상태 재탭은 취소다. 생성 폼의 날짜는 캘린더 선택일을 그대로 쓴다(M2에 DatePicker가
/// 없는 Compose와 달리 iOS는 시각 입력에 네이티브 DatePicker를 쓰지만, 날짜 자체는 여전히 선택일 고정).
final class GroupEventsViewModel: MviViewModel {
    typealias Event = Never

    @Published private(set) var uiState: UiState

    let groupId: Int64

    private let getGroupEventsUseCase: GetGroupEventsUseCase

    private let getEventDetailUseCase: GetEventDetailUseCase

    private let createEventUseCase: CreateEventUseCase

    private let deleteEventUseCase: DeleteEventUseCase

    private let rsvpEventUseCase: RsvpEventUseCase

    private let cancelEventRsvpUseCase: CancelEventRsvpUseCase

    // 진행 중인 월 조회 Task — 월 이동을 연타해도 이전 요청을 취소해 늦게 도착한 이전 달 응답이
    // 새 달 상태를 덮어쓰지 않게 한다(Kotlin loadJob?.cancel() 레이스 픽스 미러, Finding 1)
    private var loadTask: Task<Void, Never>?

    // 서버 ISO(소수부 자릿수 초과 가능) → Date. TimeFormats.swift엔 재사용 가능한 공개 파서가
    // 없어(relative()가 내부 private 헬퍼로만 씀) 같은 소수부 스트립 방식을 여기 private로 복사한다
    private static let isoFormatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter
    }()

    static func parseIso(_ iso: String) -> Date? {
        let secondsOnly = iso.replacingOccurrences(of: "\\.\\d+", with: "", options: .regularExpression)
        return isoFormatter.date(from: secondsOnly)
    }

    static func dateKey(_ date: Date) -> String {
        let parts = Calendar.current.dateComponents([.year, .month, .day], from: date)
        return dateKey(year: parts.year ?? 0, month: parts.month ?? 0, day: parts.day ?? 0)
    }

    /// yyyy-MM-dd 키 조립 — Kotlin dateKeyOf 미러
    static func dateKey(year: Int, month: Int, day: Int) -> String {
        String(format: "%04d-%02d-%02d", year, month, day)
    }

    /// 로컬 연월일시분 → 서버로 보낼 ISO — Kotlin localToIso 미러
    static func localToIso(year: Int, month: Int, day: Int, hour: Int, minute: Int) -> String {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = day
        components.hour = hour
        components.minute = minute
        let date = Calendar.current.date(from: components) ?? Date(timeIntervalSince1970: 0)
        return isoFormatter.string(from: date)
    }

    /// 해당 월 1일의 요일(일=0…토=6) — 캘린더 앞쪽 빈 칸 수(Kotlin firstDayOfWeekOfMonth 미러).
    /// Foundation의 .weekday 컴포넌트는 캘린더 firstWeekday 설정과 무관하게 1=일요일 고정이라
    /// java.time DayOfWeek(월=1…일=7)처럼 별도 모듈러 변환이 필요 없다
    static func firstDayOfWeekOfMonth(year: Int, month: Int) -> Int {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = 1
        guard let date = Calendar.current.date(from: components) else { return 0 }
        return Calendar.current.component(.weekday, from: date) - 1
    }

    /// 해당 일자의 요일(일=0…토=6) — 월 1일 요일에서 상대 계산(Kotlin dayOfWeek 미러)
    static func dayOfWeek(year: Int, month: Int, day: Int) -> Int {
        (firstDayOfWeekOfMonth(year: year, month: month) + day - 1) % 7
    }

    /// 해당 월의 일수(28~31) — Kotlin monthLength 미러
    static func monthLength(year: Int, month: Int) -> Int {
        var components = DateComponents()
        components.year = year
        components.month = month
        components.day = 1
        guard let date = Calendar.current.date(from: components),
              let range = Calendar.current.range(of: .day, in: .month, for: date) else { return 30 }
        return range.count
    }

    func onAction(_ action: Action) {
        switch action {
        case .refresh: loadMonth()
        case .moveMonth(let delta): moveMonth(delta)
        case .goToday: goToday()
        case .selectDay(let dateKey): uiState.selectedDay = dateKey
        case .toggleCreateForm:
            uiState.showCreateForm.toggle()
            uiState.createError = nil
        case .createEvent(let title, let location, let description, let startHour, let startMinute, let endHour, let endMinute):
            createEvent(title: title, location: location, description: description,
                        startHour: startHour, startMinute: startMinute, endHour: endHour, endMinute: endMinute)
        case .deleteEvent(let eventId): deleteEvent(eventId)
        case .rsvp(let event, let status): rsvp(event, status)
        case .toggleAttendees(let eventId): toggleAttendees(eventId)
        case .dismissActionError: uiState.actionError = nil
        }
    }

    /// 현재 앵커 월 재조회 — [월초, 다음달 초) ISO(웹 loadMonth 미러). 진행 중인 이전 요청은
    /// 무조건 취소하고 교체한다(재진입 가드 없음) — 월 이동을 빠르게 연타해도 늦게 도착한 이전 달
    /// 응답이 새 달 상태를 덮어쓰지 못하게. 취소된 Task는 응답 수신 후·에러 처리 전 각각
    /// Task.isCancelled로 걸러 상태를 쓰지 않는다(Kotlin CancellationException rethrow 미러)
    private func loadMonth() {
        loadTask?.cancel()

        uiState.isLoading = true
        uiState.error = nil
        let year = uiState.year
        let month = uiState.month
        let (nextYear, nextMonth) = month == 12 ? (year + 1, 1) : (year, month + 1)

        loadTask = Task { @MainActor in
            do {
                let events = try await getGroupEventsUseCase.invoke(
                    groupId: groupId,
                    fromIso: Self.localToIso(year: year, month: month, day: 1, hour: 0, minute: 0),
                    toIso: Self.localToIso(year: nextYear, month: nextMonth, day: 1, hour: 0, minute: 0)
                )
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.events = events
            } catch {
                if Task.isCancelled { return }
                uiState.isLoading = false
                uiState.error = error.kotlinMessage(fallback: "일정을 불러오지 못했습니다.")
            }
        }
    }

    /// 월 이동 — 이동한 달의 1일을 선택해 아래 목록이 이전 달 잔상을 가리키지 않게 한다(웹 미러).
    /// isLoading은 손대지 않는다 — loadMonth()가 무조건 취소+재시작하므로 재진입 가드용 수동
    /// 리셋이 필요 없다(레이스 픽스로 재진입 가드 자체가 사라졌다)
    private func moveMonth(_ delta: Int) {
        let zeroBased = uiState.year * 12 + (uiState.month - 1) + delta
        let year = zeroBased / 12
        let month = zeroBased % 12 + 1

        uiState.year = year
        uiState.month = month
        uiState.selectedDay = Self.dateKey(year: year, month: month, day: 1)
        uiState.events = nil
        loadMonth()
    }

    /// 오늘로 — 다른 달이면 재조회, 같은 달이면 선택만 이동(웹 goToday 미러)
    private func goToday() {
        let today = Calendar.current.dateComponents([.year, .month, .day], from: Date())
        guard let year = today.year, let month = today.month, let day = today.day else { return }
        let key = Self.dateKey(year: year, month: month, day: day)

        if year != uiState.year || month != uiState.month {
            uiState.year = year
            uiState.month = month
            uiState.selectedDay = key
            uiState.events = nil
            loadMonth()
        } else {
            uiState.selectedDay = key
        }
    }

    /// 생성 — 날짜는 캘린더 선택일 고정이라 결과는 항상 현재 달(목록 삽입+정렬만)
    private func createEvent(
        title: String, location: String, description: String,
        startHour: Int, startMinute: Int, endHour: Int?, endMinute: Int?
    ) {
        if uiState.isCreating { return }

        let parts = uiState.selectedDay.split(separator: "-").compactMap { Int($0) }
        guard parts.count == 3 else { return }
        var endsAtIso: String? = nil
        if let endHour = endHour, let endMinute = endMinute {
            // 종료가 시작보다 빠르면 폼 검증에서 걸렀어야 한다 — 방어적으로 한 번 더(Kotlin 미러)
            if endHour < startHour || (endHour == startHour && endMinute < startMinute) {
                uiState.createError = "종료 시각은 시작 시각보다 빠를 수 없습니다"
                return
            }
            endsAtIso = Self.localToIso(year: parts[0], month: parts[1], day: parts[2], hour: endHour, minute: endMinute)
        }

        uiState.isCreating = true
        uiState.createError = nil
        Task { @MainActor in
            do {
                let created = try await createEventUseCase.invoke(
                    groupId: groupId,
                    title: title,
                    description: description.isEmpty ? nil : description,
                    location: location.isEmpty ? nil : location,
                    startsAtIso: Self.localToIso(year: parts[0], month: parts[1], day: parts[2], hour: startHour, minute: startMinute),
                    endsAtIso: endsAtIso
                )
                uiState.isCreating = false
                uiState.showCreateForm = false
                uiState.events = ((uiState.events ?? []) + [created]).sorted { $0.startsAt < $1.startsAt }
            } catch {
                uiState.isCreating = false
                uiState.createError = error.kotlinMessage(fallback: "일정을 만들지 못했습니다.")
            }
        }
    }

    /// 삭제(작성자/모더레이터 — 서버 검증) — 성공 시 목록에서 제거(웹 handleDelete 미러)
    private func deleteEvent(_ eventId: Int64) {
        if uiState.busyEventId != nil { return }

        uiState.busyEventId = eventId
        uiState.actionError = nil
        Task { @MainActor in
            do {
                try await deleteEventUseCase.invoke(groupId: groupId, eventId: eventId)
                uiState.busyEventId = nil
                uiState.events = uiState.events?.filter { $0.id != eventId }
            } catch {
                uiState.busyEventId = nil
                uiState.actionError = error.kotlinMessage(fallback: "일정 삭제에 실패했습니다.")
            }
        }
    }

    /// RSVP — 같은 상태 재탭=취소(웹 handleRsvp 미러). 서버가 돌려준 집계 갱신 일정으로 카드를 교체한다
    private func rsvp(_ event: GroupEvent, _ status: RsvpStatus) {
        if uiState.busyEventId != nil { return }

        uiState.busyEventId = event.id
        uiState.actionError = nil
        Task { @MainActor in
            do {
                let updated = event.myRsvp == status
                    ? try await cancelEventRsvpUseCase.invoke(groupId: groupId, eventId: event.id)
                    : try await rsvpEventUseCase.invoke(groupId: groupId, eventId: event.id, status: status)
                uiState.busyEventId = nil
                uiState.events = uiState.events?.map { $0.id == updated.id ? updated : $0 }
                // 명단이 바뀌었으니 펼쳐볼 때 다시 불러온다(웹 setAttendees(null) 미러)
                uiState.attendeesByEvent.removeValue(forKey: event.id)
            } catch {
                uiState.busyEventId = nil
                uiState.actionError = error.kotlinMessage(fallback: "참석 응답에 실패했습니다.")
            }
        }
    }

    /// 참석자 펼침 — 캐시 없으면 단건 조회, 실패는 조용히 빈 목록(부가 정보 — 웹 미러)
    private func toggleAttendees(_ eventId: Int64) {
        if uiState.expandedEventIds.contains(eventId) {
            uiState.expandedEventIds.remove(eventId)
            return
        }
        uiState.expandedEventIds.insert(eventId)
        if uiState.attendeesByEvent[eventId] == nil {
            Task { @MainActor in
                let attendees = (try? await getEventDetailUseCase.invoke(groupId: groupId, eventId: eventId).attendees) ?? []
                uiState.attendeesByEvent[eventId] = attendees
            }
        }
    }

    init(
        groupId: Int64,
        getGroupEventsUseCase: GetGroupEventsUseCase = AppContainer.shared.getGroupEventsUseCase,
        getEventDetailUseCase: GetEventDetailUseCase = AppContainer.shared.getEventDetailUseCase,
        createEventUseCase: CreateEventUseCase = AppContainer.shared.createEventUseCase,
        deleteEventUseCase: DeleteEventUseCase = AppContainer.shared.deleteEventUseCase,
        rsvpEventUseCase: RsvpEventUseCase = AppContainer.shared.rsvpEventUseCase,
        cancelEventRsvpUseCase: CancelEventRsvpUseCase = AppContainer.shared.cancelEventRsvpUseCase,
        getCurrentUserIdUseCase: GetCurrentUserIdUseCase = AppContainer.shared.getCurrentUserIdUseCase
    ) {
        self.groupId = groupId
        self.getGroupEventsUseCase = getGroupEventsUseCase
        self.getEventDetailUseCase = getEventDetailUseCase
        self.createEventUseCase = createEventUseCase
        self.deleteEventUseCase = deleteEventUseCase
        self.rsvpEventUseCase = rsvpEventUseCase
        self.cancelEventRsvpUseCase = cancelEventRsvpUseCase
        let today = Calendar.current.dateComponents([.year, .month, .day], from: Date())
        uiState = UiState(
            myUserId: getCurrentUserIdUseCase.invoke()?.int64Value,
            year: today.year ?? 0,
            month: today.month ?? 0,
            selectedDay: Self.dateKey(year: today.year ?? 0, month: today.month ?? 0, day: today.day ?? 0)
        )
        loadMonth()
    }

    struct UiState {
        // 일정 카드의 "작성자 본인" 삭제 판정 — 세션이 있는 한 nil이 아니다
        var myUserId: Int64? = nil
        // 월 앵커(연/월) — 범위 조회와 캘린더 그리드의 기준
        var year: Int = 0
        var month: Int = 0
        // 선택일 키(yyyy-MM-dd, 로컬) — 셀 강조와 아래 목록·생성 폼 날짜의 기준
        var selectedDay: String = ""
        // nil=이 달을 아직 못 읽음(로딩/실패) — 빈 목록과 구분한다(웹 events === null 미러)
        var events: [GroupEvent]? = nil
        var isLoading = false
        var error: String? = nil
        var showCreateForm = false
        var isCreating = false
        var createError: String? = nil
        // 카드 액션(RSVP/삭제) 진행 중 일정 — 동시에 하나만(멤버 탭 processingRequestUserId 선례)
        var busyEventId: Int64? = nil
        // RSVP/삭제 실패 문구 — 로드 에러(error)와 달리 캘린더를 대체하지 않는다
        var actionError: String? = nil
        var expandedEventIds: Set<Int64> = []
        // 일정별 참석자 캐시 — RSVP가 바뀌면 그 일정 키를 비워 다음 펼침에 재조회한다
        var attendeesByEvent: [Int64: [EventAttendee]] = [:]

        /// 로컬 날짜 키별 일정 — 캘린더 점·선택일 목록의 원천(웹 eventsByDay 미러). 접근마다
        /// groupBy를 다시 도는 계산 프로퍼티라 — 소비 측(GroupEventsTab)이 1회만 읽어 재사용해야
        /// 한다(Kotlin Finding 2 레이스 픽스 미러, 여기 자체는 손대지 않는다)
        var eventsByDay: [String: [GroupEvent]] {
            Dictionary(grouping: events ?? []) { event in
                GroupEventsViewModel.parseIso(event.startsAt).map(GroupEventsViewModel.dateKey)
                    ?? String(event.startsAt.prefix(10))
            }
        }

        var selectedEvents: [GroupEvent] { eventsByDay[selectedDay] ?? [] }
    }

    enum Action {
        case refresh
        case moveMonth(Int)
        case goToday
        case selectDay(String)
        case toggleCreateForm
        case createEvent(title: String, location: String, description: String,
                         startHour: Int, startMinute: Int, endHour: Int?, endMinute: Int?)
        case deleteEvent(Int64)
        case rsvp(GroupEvent, RsvpStatus)
        case toggleAttendees(Int64)
        case dismissActionError
    }
}

File: composeApp/src/commonMain/kotlin/kr/hhp227/storygroup/ui/screens/group/GroupEventsViewModel.kt
package kr.hhp227.storygroup.ui.screens.group

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.emptyFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kr.hhp227.storygroup.shared.domain.model.EventAttendee
import kr.hhp227.storygroup.shared.domain.model.GroupEvent
import kr.hhp227.storygroup.shared.domain.model.RsvpStatus
import kr.hhp227.storygroup.shared.domain.usecase.CancelEventRsvpUseCase
import kr.hhp227.storygroup.shared.domain.usecase.CreateEventUseCase
import kr.hhp227.storygroup.shared.domain.usecase.DeleteEventUseCase
import kr.hhp227.storygroup.shared.domain.usecase.GetCurrentUserIdUseCase
import kr.hhp227.storygroup.shared.domain.usecase.GetEventDetailUseCase
import kr.hhp227.storygroup.shared.domain.usecase.GetGroupEventsUseCase
import kr.hhp227.storygroup.shared.domain.usecase.RsvpEventUseCase
import kr.hhp227.storygroup.ui.mvi.MviViewModel
import kr.hhp227.storygroup.ui.util.dateKeyOf
import kr.hhp227.storygroup.ui.util.isoToLocal
import kr.hhp227.storygroup.ui.util.localToIso
import kr.hhp227.storygroup.ui.util.todayLocal
import org.jetbrains.compose.resources.getString
import storygroup.composeapp.generated.resources.Res
import storygroup.composeapp.generated.resources.events_error_create
import storygroup.composeapp.generated.resources.events_error_delete
import storygroup.composeapp.generated.resources.events_error_end_before_start
import storygroup.composeapp.generated.resources.events_error_load
import storygroup.composeapp.generated.resources.events_error_rsvp

/**
 * 일정 탭 — 웹 /groups/[id]/events 캘린더 페이지 미러(탭별 VM 분리, 레거시 탭 Fragment VM 구조).
 * 월 앵커 기준 [월초, 다음달 초) 범위를 조회하고, 날짜 귀속·선택 키는 기기 로컬 타임존
 * yyyy-MM-dd(웹 ymd 미러). RSVP는 낙관적 갱신 없이 서버가 돌려준 집계 갱신 일정으로 카드를
 * 교체하고, 같은 상태 재탭은 취소다. 생성 폼의 날짜는 캘린더 선택일을 그대로 쓴다(M2에
 * DatePicker가 없음) — 그래서 새 일정은 항상 보고 있는 달에 속하고, 웹의 "다른 달이면 이동"
 * 분기가 필요 없다. iosApp GroupEventsViewModel.swift와 1:1 미러
 */
class GroupEventsViewModel(
    val groupId: Long,
    private val getGroupEventsUseCase: GetGroupEventsUseCase,
    private val getEventDetailUseCase: GetEventDetailUseCase,
    private val createEventUseCase: CreateEventUseCase,
    private val deleteEventUseCase: DeleteEventUseCase,
    private val rsvpEventUseCase: RsvpEventUseCase,
    private val cancelEventRsvpUseCase: CancelEventRsvpUseCase,
    getCurrentUserIdUseCase: GetCurrentUserIdUseCase
) : ViewModel(), MviViewModel<GroupEventsViewModel.UiState, GroupEventsViewModel.Action, Nothing> {
    private val _uiState: MutableStateFlow<UiState>
    override val uiState: StateFlow<UiState>

    override val event: Flow<Nothing> = emptyFlow()

    // 진행 중인 월 조회 Job — 월 이동을 연타해도 이전 요청을 취소해 늦게 도착한 이전 달
    // 응답이 새 달 상태를 덮어쓰지 않게 한다(Finding 1 레이스 픽스)
    private var loadJob: Job? = null

    init {
        val today = todayLocal()

        _uiState = MutableStateFlow(
            UiState(
                myUserId = getCurrentUserIdUseCase(),
                year = today.year,
                month = today.month,
                selectedDay = today.dateKey
            )
        )
        uiState = _uiState.asStateFlow()
        loadMonth()
    }

    override fun onAction(action: Action) {
        when (action) {
            Action.Refresh -> loadMonth()
            is Action.MoveMonth -> moveMonth(action.delta)
            Action.GoToday -> goToday()
            is Action.SelectDay -> _uiState.update { it.copy(selectedDay = action.dateKey) }
            Action.ToggleCreateForm -> _uiState.update {
                it.copy(showCreateForm = !it.showCreateForm, createError = null)
            }
            is Action.CreateEvent -> createEvent(action)
            is Action.DeleteEvent -> deleteEvent(action.eventId)
            is Action.Rsvp -> rsvp(action.event, action.status)
            is Action.ToggleAttendees -> toggleAttendees(action.eventId)
            Action.DismissActionError -> _uiState.update { it.copy(actionError = null) }
        }
    }

    /**
     * 현재 앵커 월 재조회 — [월초, 다음달 초) ISO(웹 loadMonth 미러). 진행 중인 이전 요청은
     * 취소하고 교체한다(Finding 1) — 월 이동을 빠르게 연타해도 늦게 도착한 이전 달 응답이
     * 새 달 상태를 덮어쓰지 못하게. runCatching이 CancellationException까지 삼키므로
     * onFailure에서 명시적으로 rethrow해 취소를 정상 전파한다(프로젝트 규칙)
     */
    private fun loadMonth() {
        loadJob?.cancel()

        _uiState.update { it.copy(isLoading = true, error = null) }
        loadJob = viewModelScope.launch {
            val state = _uiState.value
            val (nextYear, nextMonth) =
                if (state.month == 12) state.year + 1 to 1 else state.year to state.month + 1

            runCatching {
                getGroupEventsUseCase(
                    groupId,
                    localToIso(state.year, state.month, 1, 0, 0),
                    localToIso(nextYear, nextMonth, 1, 0, 0)
                )
            }.onSuccess { events ->
                _uiState.update { it.copy(isLoading = false, events = events) }
            }.onFailure { e ->
                if (e is CancellationException) throw e
                _uiState.update { it.copy(isLoading = false, error = e.message ?: getString(Res.string.events_error_load)) }
            }
        }
    }

    /** 월 이동 — 이동한 달의 1일을 선택해 아래 목록이 이전 달 잔상을 가리키지 않게 한다(웹 미러) */
    private fun moveMonth(delta: Int) {
        val state = _uiState.value
        // -1/+1만 쓰지만 산술은 일반화해 둔다(0~11 순환)
        val zeroBased = state.year * 12 + (state.month - 1) + delta
        val year = zeroBased / 12
        val month = zeroBased % 12 + 1

        _uiState.update {
            it.copy(year = year, month = month, selectedDay = dateKeyOf(year, month, 1), events = null)
        }
        loadMonth()
    }

    /** 오늘로 — 다른 달이면 재조회, 같은 달이면 선택만 이동(웹 goToday 미러) */
    private fun goToday() {
        val today = todayLocal()
        val state = _uiState.value

        if (today.year != state.year || today.month != state.month) {
            _uiState.update {
                it.copy(year = today.year, month = today.month, selectedDay = today.dateKey, events = null)
            }
            loadMonth()
        } else {
            _uiState.update { it.copy(selectedDay = today.dateKey) }
        }
    }

    /** 생성 — 날짜는 캘린더 선택일 고정이라 결과는 항상 현재 달: 목록에 넣고 정렬만 하면 된다 */
    private fun createEvent(action: Action.CreateEvent) {
        if (_uiState.value.isCreating) return

        val state = _uiState.value
        val (year, month, day) = state.selectedDay.split("-").map { it.toInt() }
        val endsAtIso = if (action.endHour != null && action.endMinute != null) {
            // 종료가 시작보다 빠르면 폼 검증에서 걸렀어야 한다 — 방어적으로 한 번 더
            if (action.endHour < action.startHour ||
                (action.endHour == action.startHour && action.endMinute < action.startMinute)
            ) {
                // getString은 suspend — 동기 검증 흐름은 그대로 두고 문구 해석만 코루틴에서
                viewModelScope.launch {
                    _uiState.update { it.copy(createError = getString(Res.string.events_error_end_before_start)) }
                }
                return
            }
            localToIso(year, month, day, action.endHour, action.endMinute)
        } else null

        _uiState.update { it.copy(isCreating = true, createError = null) }
        viewModelScope.launch {
            runCatching {
                createEventUseCase(
                    groupId = groupId,
                    title = action.title,
                    description = action.description.ifBlank { null },
                    location = action.location.ifBlank { null },
                    startsAtIso = localToIso(year, month, day, action.startHour, action.startMinute),
                    endsAtIso = endsAtIso
                )
            }.onSuccess { created ->
                _uiState.update {
                    it.copy(
                        isCreating = false,
                        showCreateForm = false,
                        events = ((it.events ?: emptyList()) + created).sortedBy(GroupEvent::startsAt)
                    )
                }
            }.onFailure { e ->
                _uiState.update { it.copy(isCreating = false, createError = e.message ?: getString(Res.string.events_error_create)) }
            }
        }
    }

    /** 삭제(작성자/모더레이터 — 서버 검증) — 성공 시 목록에서 제거(웹 handleDelete 미러) */
    private fun deleteEvent(eventId: Long) {
        if (_uiState.value.busyEventId != null) return

        _uiState.update { it.copy(busyEventId = eventId, actionError = null) }
        viewModelScope.launch {
            runCatching { deleteEventUseCase(groupId, eventId) }
                .onSuccess {
                    _uiState.update { state ->
                        state.copy(busyEventId = null, events = state.events?.filterNot { it.id == eventId })
                    }
                }
                .onFailure { e ->
                    _uiState.update { it.copy(busyEventId = null, actionError = e.message ?: getString(Res.string.events_error_delete)) }
                }
        }
    }

    /** RSVP — 같은 상태 재탭=취소. 서버가 돌려준 집계 갱신 일정으로 교체(웹 handleRsvp 미러) */
    private fun rsvp(event: GroupEvent, status: RsvpStatus) {
        if (_uiState.value.busyEventId != null) return

        _uiState.update { it.copy(busyEventId = event.id, actionError = null) }
        viewModelScope.launch {
            runCatching {
                if (event.myRsvp == status) cancelEventRsvpUseCase(groupId, event.id)
                else rsvpEventUseCase(groupId, event.id, status)
            }.onSuccess { updated ->
                _uiState.update { state ->
                    state.copy(
                        busyEventId = null,
                        events = state.events?.map { if (it.id == updated.id) updated else it },
                        // 명단이 바뀌었으니 펼쳐볼 때 다시 불러온다(웹 setAttendees(null) 미러)
                        attendeesByEvent = state.attendeesByEvent - event.id
                    )
                }
            }.onFailure { e ->
                _uiState.update { it.copy(busyEventId = null, actionError = e.message ?: getString(Res.string.events_error_rsvp)) }
            }
        }
    }

    /** 참석자 펼침 — 캐시 없으면 단건 조회, 실패는 조용히 빈 목록(부가 정보 — 웹 미러) */
    private fun toggleAttendees(eventId: Long) {
        val state = _uiState.value

        if (eventId in state.expandedEventIds) {
            _uiState.update { it.copy(expandedEventIds = it.expandedEventIds - eventId) }
            return
        }
        _uiState.update { it.copy(expandedEventIds = it.expandedEventIds + eventId) }
        if (eventId !in state.attendeesByEvent) {
            viewModelScope.launch {
                val attendees = runCatching { getEventDetailUseCase(groupId, eventId).attendees }
                    .getOrDefault(emptyList())

                _uiState.update { it.copy(attendeesByEvent = it.attendeesByEvent + (eventId to attendees)) }
            }
        }
    }

    data class UiState(
        // 일정 카드의 "작성자 본인" 삭제 판정 — 세션이 있는 한 null이 아니다
        val myUserId: Long? = null,
        // 월 앵커(연/월) — 범위 조회와 캘린더 그리드의 기준
        val year: Int = 0,
        val month: Int = 0,
        // 선택일 키(yyyy-MM-dd, 로컬) — 셀 강조와 아래 목록·생성 폼 날짜의 기준
        val selectedDay: String = "",
        // null=이 달을 아직 못 읽음(로딩/실패) — 빈 목록과 구분한다(웹 events === null 미러)
        val events: List<GroupEvent>? = null,
        val isLoading: Boolean = false,
        val error: String? = null,
        val showCreateForm: Boolean = false,
        val isCreating: Boolean = false,
        val createError: String? = null,
        // 카드 액션(RSVP/삭제) 진행 중 일정 — 동시에 하나만(멤버 탭 processingRequestUserId 선례)
        val busyEventId: Long? = null,
        // RSVP/삭제 실패 문구 — 로드 에러(error)와 달리 캘린더를 대체하지 않는다
        val actionError: String? = null,
        val expandedEventIds: Set<Long> = emptySet(),
        // 일정별 참석자 캐시 — RSVP가 바뀌면 그 일정 키를 비워 다음 펼침에 재조회한다
        val attendeesByEvent: Map<Long, List<EventAttendee>> = emptyMap()
    ) {
        /** 로컬 날짜 키별 일정 — 캘린더 점·선택일 목록의 원천(웹 eventsByDay 미러) */
        val eventsByDay: Map<String, List<GroupEvent>>
            get() = (events ?: emptyList()).groupBy { event ->
                isoToLocal(event.startsAt)?.dateKey ?: event.startsAt.take(10)
            }

        val selectedEvents: List<GroupEvent> get() = eventsByDay[selectedDay] ?: emptyList()
    }

    sealed interface Action {
        data object Refresh : Action
        data class MoveMonth(val delta: Int) : Action
        data object GoToday : Action
        data class SelectDay(val dateKey: String) : Action
        data object ToggleCreateForm : Action
        data class CreateEvent(
            val title: String,
            val location: String,
            val description: String,
            val startHour: Int,
            val startMinute: Int,
            val endHour: Int?,
            val endMinute: Int?
        ) : Action
        data class DeleteEvent(val eventId: Long) : Action
        data class Rsvp(val event: GroupEvent, val status: RsvpStatus) : Action
        data class ToggleAttendees(val eventId: Long) : Action
        data object DismissActionError : Action
    }
}
